apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/AllocateAndAddFloatingIpToNode.java

```diff
@@ -225,28 +228,64 @@ public boolean apply(@Nullable org.jclouds.openstack.neutron.v2.domain.FloatingI
+            break;
+         } catch (Exception ex) {
+            logger.trace("<< [%s] failed to allocate a floating IP from network %s for node(%s)", ex.getMessage(), network, node);
          }
-      });
-      if (!networkOptional.isPresent())
-         throw new InsufficientResourcesException("Failed to find a suitable external network.");
+      }
 
-      org.jclouds.openstack.neutron.v2.domain.FloatingIP createFloatingIP = org.jclouds.openstack.neutron.v2.domain.FloatingIP.CreateFloatingIP
-              .createBuilder(networkOptional.get().getId())
-              .availabilityZone(networkOptional.get().getAvailabilityZone())
-              .build();
+      // 2.) Attempt to allocate from a external network
+      if (floatingIP == null) {
+         Optional<Network> networkOptional = Iterables.tryFind(networks, new Predicate<Network>() {
+            @Override
+            public boolean apply(Network input) {
+               return availabilityZone.equals(input.getAvailabilityZone()) && input.getExternal();
+            }
+         });
+         if (networkOptional.isPresent()) {
+            org.jclouds.openstack.neutron.v2.domain.FloatingIP createFloatingIP = org.jclouds.openstack.neutron.v2.domain.FloatingIP.CreateFloatingIP
+                    .createBuilder(networkOptional.get().getId())
+                    .availabilityZone(networkOptional.get().getAvailabilityZone())
+                    .build();
+            floatingIP = neutronFloatingApi.create((org.jclouds.openstack.neutron.v2.domain.FloatingIP.CreateFloatingIP) createFloatingIP);
+            logger.debug(">> allocated floating IP(%s) from network(%s) for node(%s)", floatingIP, networkOptional.get(), node);
+         }
+      }
 
-      org.jclouds.openstack.neutron.v2.domain.FloatingIP floatingIP = neutronFloatingApi.create((org.jclouds.openstack.neutron.v2.domain.FloatingIP.CreateFloatingIP) createFloatingIP);
-      floatingIpCache.asMap().putIfAbsent(RegionAndId.fromSlashEncoded(node.getId()), ImmutableList.of(FloatingIpForServer.create(RegionAndId.fromSlashEncoded(node.getId()), floatingIP.getId(), floatingIP.getFloatingIpAddress())));
+      if (floatingIP == null) {
+         throw new IllegalStateException("Failed to allocate a floating IP for node " + node + ".\n" +
+                 "Failed to find suitable external networks or to allocate from poolNames specified: "
+                 + Iterables.toString(poolNames.get()));
+      }
+      floatingIpCache.asMap().put(RegionAndId.fromSlashEncoded(node.getId()), ImmutableList.of(FloatingIpForServer.create(RegionAndId.fromSlashEncoded(node.getId()), floatingIP.getId(), floatingIP.getFloatingIpAddress())));
       return floatingIP;
    }
 
+
+   private List<Network> getNetworksFromPoolNames(String regionId, final Optional<Set<String>> poolNames) {
+      return getNetworkApi(regionId).list().concat().filter(new Predicate<Network>() {
```

Comment:
if the pool name set is absent or empty, just return an empty list here. There is no need to filter in that case, nor to have that check in the predicate, applying it on every iteration.